repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/RequestHandler.scala

```diff
@@ -0,0 +1,524 @@
+          case SC_BAD_REQUEST => Failure(IdentityValidationException("Bad Token Validation request to identity!"))
+          case SC_UNAUTHORIZED => Failure(AdminTokenUnauthorizedException("Unable to validate token, authenticating token unauthorized"))
+          case SC_FORBIDDEN => Failure(IdentityAdminTokenException("Admin token unauthorized to validate token"))
+          case SC_NOT_FOUND =>
+            Option(config.getCache) foreach { cacheSettings =>
+              // never null because configurationUpdated in KeystoneV2Filter
+              val timeout = offsetTtl(cacheSettings.getTimeouts.getToken, cacheSettings.getTimeouts.getVariability)
+              datastore.put(s"$TOKEN_KEY_PREFIX$token", InvalidToken, timeout, TimeUnit.SECONDS)
+            }
+            Success(InvalidToken)
+          case SC_SERVICE_UNAVAILABLE => Failure(IdentityValidationException("Identity Service not available to authenticate token"))
+          case _ => Failure(IdentityCommuncationException("Unhandled response from Identity, unable to continue"))
+        }
+      case Failure(x) => Failure(IdentityCommuncationException("Unable to successfully validate token with Identity", x))
+    }
+  }
+
+  /**
+   * Check the cache, or call to identity to get the admin token
+   * @return Returns a Successful token, or a Failure
+   */
+  def getAdminToken: Try[String] = {
+    //Check the cache first, then try the request
+    Option(datastore.get(ADMIN_TOKEN_KEY)) map { value =>
+      Success(value.asInstanceOf[String])
+    } getOrElse {
+      //authenticate, or get the admin token
+      val identityEndpoint = config.getIdentityService.getUri
+
+      import play.api.libs.json._
+      val adminUsername = config.getIdentityService.getUsername
+      val adminPassword = config.getIdentityService.getPassword
+
+      val authenticationPayload = Json.obj(
+        "auth" -> Json.obj(
+          "passwordCredentials" -> Json.obj(
+            "username" -> adminUsername,
+            "password" -> adminPassword
+          )
+        )
+      )
+
+      val akkaResponse = Try(akkaServiceClient.post(ADMIN_TOKEN_KEY,
+        s"$identityEndpoint$TOKEN_ENDPOINT",
+        (Map(CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON)
+          ++ traceId.map(CommonHttpHeader.TRACE_GUID.toString -> _)).asJava,
+        Json.stringify(authenticationPayload),
+        MediaType.APPLICATION_JSON_TYPE
+      ))
+
+      akkaResponse match {
+        case Success(x) =>
+          val jsonResponse = Source.fromInputStream(x.getData).getLines().mkString("")
+          val json = Json.parse(jsonResponse)
+          Try(Success((json \ "access" \ "token" \ "id").as[String])) match {
+            case Success(s) =>
+              datastore.put(ADMIN_TOKEN_KEY, s.get)
+              s
+            case Failure(f) => Failure(IdentityCommuncationException("Token not found in identity response during Admin Authentication", f))
```

Comment:
I took the IdentityCommunicationException to be failure to communicate with Identity. That is, either the Identity service is down, or it's breaking a contract. This is the latter case, where the contract is being broken since we can't extract the admin token id from the response payload.